### ISSUE TITLE
Expand hook system for arbitrary events (e.g. Drush cache-clear types)

### ISSUE THREAD

**Issue by weitzman** [consolidation/annotated-command#56]
I'm looking at porting cache-\* commands. How do you suggest we implement the feature of cache-clear where other commandfiles can alter in their own cache types. For example, views alters in the views cache as a selection during cache-clear. Would be nice to do away with the old Drush hook system for this.

**Comment by weitzman** [comment]
@greg-1-anderson Any thoughts here?

**Comment by greg-1-anderson** [comment]
What if we did it like this.

In any commandfile:
```
/**
 * @hook on-event cache-clear
 */
public function viewsCacheClearEvent(/* arbitrary parameters, as defined by cache-clear */)
{
    // do the needful, return what cache-clear expects
}
```
Then, to utilize this in the cache-clear command:
```
class CacheClearCommand implements CustomEventAwareInterface
{
    use CustomEventAwareTrait;

    /**
     * @command cache-clear
     */
    public cacheClear($clearWhat, $options = [])
    {
        $handlers = $this->getCustomEventHandlers('cache-clear');
        // iterate and call $handlers
    }
}
```
`getCustomEventHandlers` is provided by the `CustomEventAwareTrait`, and we rely on the DI container to inject the reference to the commandfiles that declare they need it.

We could also just expose the HookManager via HookManagerAwareInterface / HookManagerAwareTrait, but I think it's better to create a custom event manager object to insulate the caller from the complexity of the full hook manager.

**Comment by weitzman** [comment]
Looks great to me ... See the bottom of ViewsCommand for commented out code where this can be used.

**Comment by greg-1-anderson** [comment]
Implemented in #60, and added to the Robo DI container in https://github.com/consolidation/Robo/pull/496
